File: app/models/Rule.java
package models;

import java.io.Serializable;
import java.util.Collections;
import java.util.Comparator;
import java.util.Date;
import java.util.List;

import javax.annotation.CheckForNull;
import javax.annotation.Nonnull;

import org.apache.commons.lang.StringUtils;
import org.apache.commons.lang.builder.EqualsBuilder;
import org.apache.commons.lang.builder.HashCodeBuilder;

import play.Logger;
import rules.RuleType;

import com.google.appengine.api.datastore.DatastoreService;
import com.google.appengine.api.datastore.DatastoreServiceFactory;
import com.google.appengine.api.datastore.Entity;
import com.google.appengine.api.datastore.FetchOptions;
import com.google.appengine.api.datastore.Key;
import com.google.appengine.api.datastore.Query;
import com.google.appengine.repackaged.com.google.common.primitives.Ints;
import com.google.common.base.Objects;
import com.google.common.base.Preconditions;
import com.google.common.collect.Lists;

/**
 * A sorting rule that allows files to be moved to a specified
 * location when certain criteria are met.
 * 
 * @author mustpax
 * @author syyang
 */
public class Rule implements Serializable {
	
    private static final long serialVersionUID = 45L;

    public static final String KIND = "Rule";
    public static final RuleMapper MAPPER = new RuleMapper();

    public static final int MAX_RULES = 200;

    private static final Comparator<Rule> RANK_COMPARATOR = new Comparator<Rule>() {
        @Override public int compare(Rule rule1, Rule rule2) {
            assert rule1 != null && rule2 != null : "rules can't be null";
            return Ints.compare(rule1.rank, rule2.rank);
        }
    };

    public Long id;
    public Long owner;
    public RuleType type;
    public String pattern;
    public String dest;
    public Integer rank;
    public Date created;
    
    public Rule() {}
    
    public Rule(RuleType type, String pattern, String dest, Integer rank, Long owner) {
        this.type = type;
        this.pattern = pattern;
        this.dest = dest;
        this.rank = rank;
        this.owner = owner;
        this.created = new Date();
    }
    
    private Rule(Entity entity) {
        this.id = entity.getKey().getId();
        this.owner = entity.getKey().getParent().getId();
        this.type = RuleType.fromDbValue((String) entity.getProperty("type"));
        this.pattern = (String) entity.getProperty("pattern");
        this.dest = (String) entity.getProperty("dest");
        this.rank = ((Long) entity.getProperty("rank")).intValue();
        this.created = (Date) entity.getProperty("created");
    }

    /** 
     * @param userId the user id
     * @return all the rules for the given user id, sorted by rank.
     * <p>
     * NOTE: the max number of rules is bound by {@link #MAX_RULES_TO_FETCH}
     */
    public static List<Rule> findByUserId(long userId) {
        @SuppressWarnings("unchecked")
        List<Rule> rules = (List<Rule>) play.cache.Cache.get(cacheKey(userId));

        if (rules == null) {
	        Query q = byOwner(userId);
	        rules = Lists.newArrayList(fetch(q));
	        Collections.sort(rules, RANK_COMPARATOR);
	        play.cache.Cache.set(cacheKey(userId), rules);
        }

        return rules;
    }
    
    public static Iterable<Rule> fetch(Query q) {
        return fetch(q, -1);
    }

    public static Iterable<Rule> fetch(Query q, int limit) {
        if (limit < 0) {
            limit = MAX_RULES;
        }

        FetchOptions fo = FetchOptions.Builder.withLimit(limit);
        return DatastoreUtil.query(q, fo, MAPPER);
    }
    
    public static Iterable<Key> fetchKeys(Query q) {
        return fetchKeys(q, -1);
    }
    
    public static Iterable<Key> fetchKeys(Query q, int limit) {
        if (limit < 0) {
            limit = MAX_RULES;
        }

        FetchOptions fo = FetchOptions.Builder.withLimit(limit);
        return DatastoreUtil.queryKeys(q, fo, MAPPER);
    }
            

    public static Query all() {
        return new Query(KIND);
    }
    
    public static Query byOwner(long userId) {
	    return all().setAncestor(User.key(userId));
    }


    public static boolean ruleExists(Long userId) {
        return fetch(byOwner(userId).setKeysOnly()).iterator().hasNext();
    }
    
    public boolean matches(String fileName) {
        return type.matches(this.pattern, fileName);
    }

    public @Nonnull List<RuleError> validate() {
        List<RuleError> ret = Lists.newLinkedList();
        if (type == null) {
            ret.add(new RuleError("type", "Missing or invalid type."));
        }

        if (owner == null) {
            ret.add(new RuleError("owner", "Missing owner."));
        }
        
        if (StringUtils.isBlank(pattern)) {
            ret.add(new RuleError("pattern", "Can't be empty."));
        } else if (pattern.contains("/")) {
            ret.add(new RuleError("pattern", "Can't contain slashes (/)."));
        }
        // Extensions may not include periods
        else if ((type == RuleType.EXT_EQ) &&
	              pattern.contains(".")) {
            ret.add(new RuleError("pattern", "Can't contain periods."));
        }

        if (StringUtils.isBlank(dest)) {
            ret.add(new RuleError("dest", "Can't be empty."));
        } else if (! dest.startsWith("/")) {
            ret.add(new RuleError("dest", "Must start with a slash (/)."));
        }
        
        // TODO check dest is not a file

        return ret;
    }

    @Override
    public int hashCode() {
        HashCodeBuilder hash = new HashCodeBuilder()
            .append(this.id)
            .append(this.type)
            .append(this.pattern)
            .append(this.dest)
            .append(this.rank)
            .append(this.owner)
            .append(this.created);
        return hash.hashCode();
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj)
            return true;
        if (getClass() != obj.getClass())
            return false;
        Rule other = (Rule) obj;
        EqualsBuilder eq = new EqualsBuilder()
            .append(this.type, other.type)
            .append(this.pattern, other.pattern)
            .append(this.dest, other.dest)
            .append(this.rank, other.rank)
            .append(this.owner, other.owner);
        return eq.isEquals();
    }
    
    @Override
    public String toString() {
        return Objects.toStringHelper(Rule.class)
            .add("id", id)
            .add("rule type", type)
            .add("pattern", pattern)
            .add("dest", dest)
            .add("rank", rank)
            .add("owner", owner)
            .add("created", created)
            .toString();
    }

    /**
     * Replace all rules for given user with a new list of rules.
     * 
     * @param allErrors if not null, validation errors are saved here
     * @return true if there were no errors and new rules were inserted
     *              inserted
     */
    public static boolean replace(User user,
                                  List<Rule> ruleList,
                                  @CheckForNull List<List<RuleError>> allErrors) {
        Preconditions.checkNotNull(ruleList);
        if (ruleList.size() > MAX_RULES) {
            throw new TooManyRulesException(String.format("Can only have max %d rules, attempting to create %d.",
                                                          MAX_RULES, ruleList.size()));
        }

        play.cache.Cache.delete(cacheKey(user.id));

        List<Rule> toSave = Lists.newArrayList();
        boolean needToRun = true;
        List<Key> oldKeys = Lists.newArrayList(fetchKeys(byOwner(user.id)));

        if (ruleList.isEmpty()) {
            Logger.info("Deleting all rules since there are no new rules to insert.");
            // No rules inserted no need to run
            needToRun = false;
        } else {
            int rank = 0;
            for (Rule rule : ruleList) {
                rule.owner = user.id;
                List<RuleError> errors = rule.validate();
                if (errors.isEmpty()) {
                    rule.rank = rank++;
                    toSave.add(rule);
                } else {
                    needToRun = false;
                }

                if (allErrors != null) {
                    allErrors.add(errors);
                }
            }

            if (!toSave.isEmpty()) {
                Logger.info("Inserting %d new rules for user.", toSave.size());
                saveAll(toSave);
            }
        }
        
        Logger.info("Deleting %d old rules.", oldKeys.size());
        // delete existing rules
        DatastoreService ds = DatastoreServiceFactory.getDatastoreService();        
        ds.delete(oldKeys);
    
        return needToRun;
    }
    
    public static Key key(long parent, long id) {
        return User.key(parent).getChild(KIND, id);
    }
    
    /**
     * We cache all rules for a user together.
     * So the cache key depends on the ownerId.
     */
    public static String cacheKey(long ownerId) {
        return User.key(ownerId) + KIND;
    }

    private static void saveAll(Iterable<Rule> rules) {
        DatastoreUtil.put(rules, MAPPER);
    }

    public static class RuleError {
        public final String field;
        public final String msg;

        public RuleError(String field, String msg) {
            this.field = field;
            this.msg = msg;
        }
    }

    private static class RuleMapper implements Mapper<Rule> {
        private RuleMapper() {}

        @Override
        public Entity toEntity(Rule r) {
            Entity entity = DatastoreUtil.newEntity(User.key(r.owner), KIND, r.id);
            entity.setProperty("type", r.type.name());
            entity.setProperty("pattern", r.pattern);
            entity.setProperty("dest", r.dest);
            entity.setProperty("rank", r.rank);
            entity.setProperty("created", new Date());
            return entity;
        }

        @Override
        public Rule toModel(Entity entity) {
            return new Rule(entity);
        }

        @Override
        public Key toKey(Rule r) {
            assert r.id != null : "Can't get key for FileMove that hasn't been persisted yet.";
            return key(r.owner, r.id);
        }

        @Override
        public Class<Rule> getType() {
            return Rule.class;
        }
    }
}

File: src/models/rule.ts
import { Entity, Model, AbstractModelService, DatastoreUtil as dutil } from './base';
import { DatastoreKey } from '@google-cloud/datastore/entity';
import { User, UserService } from './user';

import datastore from './datastore';

import joi = require('joi');

export const MAX_RULES = 200;

function globToRegex(globStr: string): RegExp {
  let replacer = /(\W)/g;
  let regexifiedStr = globStr.replace(replacer, match => {
    if (match === '*') {
      return '.*';
    }
    if (match === '?') {
      return '.?';
    }
    return '\\' + match;
  });
  return RegExp(regexifiedStr, 'gi');
}

export class Rule implements Model<RuleKey> {
  id?: RuleKey;
  type?: string;
  pattern?: string;
  dest?: string;
  rank?: number;
  created?: Date;
}

export class RuleKey {
  ownerId?: string;
  ruleId?: string;
}

export class RuleSchema extends AbstractModelService<RuleKey, Rule> {
  MAX_RULES = MAX_RULES;
  schema = {
    type: joi.string().required().only('NAME_CONTAINS', 'GLOB', 'EXT_EQ').label('Rule type'),
    pattern: joi.string().required()
      .regex(/\//, {invert: true}) // disallow slash /
      .when('type', {is: 'EXT_EQ', then: joi.string().regex(/\./, { invert: true })}) // disallow period . when type is EXT_EQ
      .label('Rule pattern')
      // .error(() => ({message: 'Pattern is required'} as any))
      ,
    rank: joi.number().required().min(0),
    dest: joi.string().required().regex(/^\//).label('Destination'),
    created: joi.date().required()
  };
  errorOverrides: { [key: string]: { [key: string]: string }} = {
    dest: {
      'string.regex.base': 'Destination must start with a slash /'
    },
    pattern: {
      'string.regex.invert.base': 'Pattern cannot contain period . when rule type is Extension Equals'
    }
  };

  kind = 'Rule';

  makeNew(ownerId: string, props?: any): Rule {
    let id = new RuleKey();
    id.ownerId = ownerId;
    let ret: Rule = new Rule();
    ret.created = new Date();
    ret.id = id;
    if (props) {
      Object.assign(ret, props);
    }
    return ret;
  }

  toEntity(r: Rule): Entity {
    if ((! r.id) || (! r.id.ownerId)) {
      throw new Error('Cannot convert rule to entity. Missing ownerId ' + r);
    }
    return super.toEntity(r);
  }


  keyFromId(id?: RuleKey) {
    if (!id || !id.ownerId) {
      throw new Error('Cannot generate key for Rule with no owner id. ' + this);
    }
    let userKey = UserService.keyFromId(id.ownerId);
    let ruleKey;
    if (id.ruleId) {
      ruleKey = datastore.key([this.kind, datastore.int(id.ruleId)]);
    } else {
      ruleKey = datastore.key([this.kind]);
    }
    return dutil.concat([userKey, ruleKey]);
  }

  idFromKey(k: DatastoreKey): RuleKey|undefined {
    if (! k) {
      throw new Error('Cannot generate a RuleKey from null key');
    }
    if (! k.parent) {
      throw new Error('Cannot generate a RuleKey from a key with no parent ' + k);
    }
    let ret = new RuleKey();
    ret.ownerId = k.parent.id || k.parent.name;
    ret.ruleId = k.id;
    return ret;
  }

  async save(rules: Rule[]): Promise<(RuleKey|undefined)[]> {
    let error = this.validate(rules);
    if (error) {
      throw error;
    }
    let entities = rules.map(t => this.toEntity(t));
    let savedEntities = await datastore.save(entities);
    let mutationResults = savedEntities[0].mutationResults as any[];
    // mr.key is only set (i.e. non-empty) if datastore generated a key for
    // us. If we specify a key/id ahead of time, then it's empty.
    return mutationResults.map(mr => {
      if (! mr.key) {
        return undefined;
      }
      let rk = new RuleKey();
      rk.ownerId = mr.key.path[0].id || mr.key.path[0].name;
      rk.ruleId = mr.key.path[1].id || mr.key.path[1].name;
      return rk;
    });
  }

  async findByOwner(ownerId: string): Promise<Rule[]> {
    let q = this.all()
      .hasAncestor(UserService.keyFromId(ownerId))
      .order('rank')
      .limit(MAX_RULES);
    return await this.query(q);
  }

  customizeError(error: joi.ValidationErrorItem): joi.ValidationErrorItem {
    if (this.errorOverrides[error.path[1]] && this.errorOverrides[error.path[1]][error.type]) {
      error.message = this.errorOverrides[error.path[1]][error.type];
    }
    return error;
  }

  matches(rule: Rule, fileName: string): boolean {
    if (! rule.pattern) {
      return false;
    }

    switch (rule.type) {
    case 'NAME_CONTAINS':
      return fileName.toLowerCase().indexOf(rule.pattern.toLowerCase()) > -1;
    case 'GLOB':
      return !! globToRegex(rule.pattern).exec(fileName);
    case 'EXT_EQ':
      let idx = fileName.lastIndexOf('.');
      // if idx is 0 the file starts with a period . so we consider it to be dotfile
      // with no extensions
      // if idx is negative there's no period . so file has no extension
      if (idx < 1) {
        return false;
      }
      // Add one to idx to skip the period .
      let ext = fileName.substr(idx + 1);
      // Handle multiple extensions in a pattern with commas
      let patternExts = rule.pattern.split(',').map(s => s.trim());
      return patternExts.some(pat => ext.toLowerCase() === pat.toLowerCase());
    }
    return false;
  }

  async createCannedRules(user: User) {
    let rules: Rule[] = [
      this.makeNew(user.id as string, {type: 'EXT_EQ', pattern: 'jpg, png, gif', dest: '/Photos', rank: 0}),
      this.makeNew(user.id as string, {type: 'NAME_CONTAINS', pattern: 'Essay', dest: '/Documents', rank: 1}),
      this.makeNew(user.id as string, {type: 'GLOB', pattern: 'Prince*.mp3', dest: '/Music/Prince', rank: 2}),
    ];
    await this.save(rules);
    return rules;
  }
}

export let RuleService = new RuleSchema();
